Smart Contract Development – Major Refactor and Feature Expansion

kaneki003 opened the pull request:
## Updates

- **Introduced Three New Auction Contracts**
  - Added Vickrey (sealed-bid),Exponential and Logarithmic Reverse Dutch auction contracts.
  - All auctions now support ERC20 token bidding, allowing for flexible payment options (including wrapped ETH if needed).
  - Enhanced logic for secure bidding, deadline management, and safe withdrawal handling.

- **Abstracted Common Auction Logic**
  - Created a new abstract `Auction` contract to centralize shared logic and enforce consistency across different auction types.
  - Standardized common events and auction type definitions.

- **Refactored Existing Codebase**
  - Migrated Linear Reverse Dutch auction to ERC20-based bidding.
  - Unified auction data structures, event naming conventions, and access control modifiers throughout the contracts.

- **Cleaned Up Legacy Code**
  - Removed old ERC20 all-pay and highest-bidder-pay contracts.

## Additional Notes

- All auctions now work with ERC20 tokens; ETH is supported through its wrapped form, based on team feedback and discussions.

---


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added a refactored AllPayAuction contract supporting ERC20 bidding tokens with explicit bid amounts and per-bidder tracking.
  - Introduced new auction contracts: EnglishAuction, VickreyAuction (sealed-bid), LinearReverseDutchAuction, ExponentialReverseDutchAuction, and LogarithmReverseDutchAuction, all supporting NFTs and ERC20 tokens.
  - Enabled flexible auction creation with detailed parameters including bidding tokens, deadlines, price decay models, and commit/reveal phases.
  - Implemented secure bidding, deadline extensions, dynamic price calculations, and robust withdrawal mechanisms for winners and auctioneers.

- **Bug Fixes**
  - Enhanced validation for auction parameters, bidding amounts, and withdrawal conditions to prevent invalid operations and multiple withdrawals.

- **Refactor**
  - Reorganized auction contracts to inherit from a common abstract Auction contract with unified event definitions and token transfer handling.
  - Replaced ETH-based bidding with ERC20 token transfers for clearer fund management.
  - Updated event names and parameters for consistency.

- **Revert**
  - Removed legacy auction contracts including AllPayAuctionERC20, HighestBidderPayAuction, and ReverseDutchAuction in favor of improved modular implementations.

- **Tests**
  - Added comprehensive test suites for new auction contracts covering creation, bidding, price decay, withdrawals, and edge cases.
  - Refactored AllPayAuction tests to support ERC20 bidding tokens and improved validation.
  - Removed tests for deprecated auction contracts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kaneki003 commented:
@ceilican , @yogesh0509 ,I have done the contract development based on the feedback taken.Kindly review.
Also to make the review easier,i have squashed all the earlier commits into a single one to help you better track the changes.

ceilican commented on contracts/AllPayAuction.sol:
```diff
@@ -1,193 +1,179 @@
+    mapping (uint256 => AuctionData) public auctions;
 
-    struct Auction {
+    struct AuctionData{
```
space missing.

yogesh0509 replied:
We can implement a linter in our package.json along with a check on github pipelines to fix these in future easily.

ceilican commented on contracts/AllPayAuction.sol:
```diff
@@ -1,193 +1,179 @@
+ * @title AllPayAuction
+ * @notice Auction contract for NFT and token auctions,where all bidders pay their bid amount but only the highest bidder wins the auction.
+ */
+contract AllPayAuction is Auction{
```
space missing before "{".

ceilican commented on contracts/EnglishAuction.sol:
```diff
@@ -0,0 +1,178 @@
+    uint256 public auctionCounter=0;
+    mapping (uint256 => AuctionData) public auctions;
+
+    struct AuctionData{
```
Can't we move this struct and the `AuctionCreated` event to the `Auction` contract or somewhere else from where they could be reused instead of duplciated?

kaneki003 replied:
Yes @ceilican , i had thought of that but the problem was all the contracts aren't using same auction struct.For eg,reverse dutch contracts uses `decayFactor` ,`startingPrice` like terms ,which are not present in other auctions.In normal auctions also we have `minBidDelta` which is of no use to other auction contracts.

Since,`AuctionCreated` event is emitting auction params,with different params we couldn't store it common for all as well.

ceilican commented on contracts/AllPayAuction.sol:
```diff
@@ -1,193 +1,179 @@
+    ) external{
+        require(bytes(name).length>0,"Name must be present");
+        require(minBidDelta>=0,"Minimum bid delta cannot be negative");
+        require(startingBid>=0,"Starting bid cannot be negative");
```
These "requires" are not necessary. Since `minBidDelta` and `startingBid` are uint256, they cannot be negative anyway.

ceilican commented on contracts/AllPayAuction.sol:
```diff
@@ -1,193 +1,179 @@
+        require(bytes(name).length>0,"Name must be present");
+        require(minBidDelta>=0,"Minimum bid delta cannot be negative");
+        require(startingBid>=0,"Starting bid cannot be negative");
+        require(duration>0,"Duration must be greater than zero seconds");
```
This is not necessary.

ceilican commented on contracts/AllPayAuction.sol:
```diff
@@ -1,193 +1,179 @@
-            imageUrl: imageUrl,
-            auctionType: auctionType,
-            itemWithdrawn: false,
+        uint256 deadline=block.timestamp+duration;
```
Generally, use spaces before and after operators such as `=`, `+`, `>=`, `==`.

ceilican commented on contracts/EnglishAuction.sol:
```diff
@@ -0,0 +1,178 @@
+        _;
+    }
+
+    function createAuction(
```
I suspect that at least part of this function's code could be moved to the `Auction` contract as well, since it tends to be teh same for all auctions.

kaneki003 replied:
I have moved modifiers and common requires of auction params while creating new auction into abstract contract along with `auctionCounter`  identifier

ceilican commented on contracts/ExponentialReverseDutchAuction.sol:
```diff
@@ -0,0 +1,199 @@
+    }
+
+    //placeBid function is not required in this auction type as the price is determined by the auctioneer and the auctioneer can withdraw the item on placing the bid only
+    function withdrawItem(uint256 auctionId,uint256 bidAmount) external validAuctionId(auctionId) {
```
We don't need `bidAmount` here, since the bidder will pay `currentPrice` anyway.

ceilican commented on contracts/ExponentialReverseDutchAuction.sol:
```diff
@@ -0,0 +1,199 @@
+        require(bidAmount>=currentPrice,"Bid amount is less than current price");
+        
+        SafeERC20.safeTransferFrom(IERC20(auction.biddingToken),msg.sender,address(this),currentPrice);
+        if(auction.auctionType == AuctionType.NFT){
```
I think we can create an auxiliary `transferItem` function in the `Auction` base contract that encapsulates the logic of checking whether the item is an NFT or amount of FT and calls the suitable transfer function.

ceilican commented on contracts/ExponentialReverseDutchAuction.sol:
```diff
@@ -0,0 +1,199 @@
+        auction.availableFunds=0;
+
+        emit fundsWithdrawn(
+            auctionId,
```
There is no need to break lines here.  Look for other places where the formatter is exaggerating in the line breaks and eliminate the unnecessary line breaks.

ceilican commented on contracts/LinearReverseDutchAuction.sol:
```diff
@@ -0,0 +1,170 @@
+        emit itemWithdrawn(auctionId,msg.sender,auction.auctionedToken,auction.auctionedTokenIdOrAmount);
+    }
+
+    function withdrawFunds(uint256 auctionId) external validAuctionId(auctionId){
```
I think the functions `withdrawFunds` from most auctions can be moved to the base `Auction` contract.

ceilican commented on contracts/LinearReverseDutchAuction.sol:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    //placeBid function is not required in this auction type as the price is determined by the auctioneer and the auctioneer can withdraw the item on placing the bid only
+    function withdrawItem(uint256 auctionId,uint256 bidAmount) external validAuctionId(auctionId) {
```
I think this function is the same for all ReverseDutchAuction types. Therefore, we can create an abstract superclass called `ReverseDutchAuction`, a subclass of `Auction`, containing this function. Then the subclasses of `ReverseDutchAuction` will just have to implement `getCurrentPrice`.

ceilican commented on contracts/LogarithmReverseDutchAuction.sol:
```diff
@@ -0,0 +1,228 @@
+    }
+
+    //Returns the logarithm base 2 of x scaled by 10^18 with a specified number of fractional bits.
+    function log2Fixed(uint256 x,uint8 fracBits) internal pure returns (uint256){
```
Isn't there a library that we could use, instead of implementing our own logarithm functions?

ceilican commented on contracts/VickreyAuction.sol:
```diff
@@ -0,0 +1,187 @@
+        );
+    }
+
+    function commitBid(uint256 auctionId,bytes32 commitment) validAuctionId(auctionId) external payable {
+        AuctionData storage auction = auctions[auctionId]; 
+        require(block.timestamp<auction.bidCommitEnd,"The commiting phase has ended!");
+        require(commitments[auctionId][msg.sender]==bytes32(0),"The sender has already commited");
+        require(msg.value > 1000000000000000, "Minimum fees not fulfilled");  //minimum fees of 0.001 ether
+        commitments[auctionId][msg.sender]=commitment; 
+    }
```
Correct. This is important.

kaneki003 commented:
Changes made :

- Improved the abstraction of boilerplate code.
-  Added test suites to verify the functionality of the smart-contracts.
-  Implemented changes to prevent re-entrancy attacks, which helps improve the security of the smart contracts.
- Introduced a new formatter to enforce & improve code style consistency.

ceilican commented on contracts/LogarithmReverseDutchAuction.sol:
```diff
@@ -1,22 +1,20 @@
-    mapping (uint256 => AuctionData) public auctions;
-
-    struct AuctionData{
+contract LogarithmReverseDutchAuction is Auction {
```
"Logarithm" => "Logarithmic", also in the file name.
